src/datarobotx/idp/custom_application_source.py

```diff
@@ -0,0 +1,316 @@
+) -> str:
+    client = dr.Client(endpoint=endpoint, token=token)  # type: ignore
+    versions = client.get(
+        f"customApplicationSources/{custom_application_source_id}/versions/"
+    ).json()["data"]
+    for version in versions:
+        if version_token in version.get("label", ""):
+            return str(version["id"])
+    raise KeyError("No matching application source version found")
+
+
+def _get_or_create_custom_application_source_version(
+    endpoint: str,
+    token: str,
+    custom_application_source_id: str,
+    runtime_parameter_values: Optional[List[Union[RuntimeParameterValue, Dict[str, str]]]] = None,
+    previous_custom_application_source_version_id: Optional[str] = None,
+    **kwargs: Any,
+) -> str:
+    client = dr.Client(token=token, endpoint=endpoint)  # type: ignore
+
+    if runtime_parameter_values is not None:
+        runtime_parameter_values_objs = [
+            RuntimeParameterValue(**param) if isinstance(param, dict) else param
+            for param in runtime_parameter_values
+        ]
+    else:
+        runtime_parameter_values_objs = []
+
+    folder_path = kwargs.pop("folder_path", None)
+
+    version_token = get_hash(
+        Path(folder_path) if folder_path else None,
+        custom_application_source_id,
+        runtime_parameter_values=runtime_parameter_values_objs,
+        **kwargs,
+    )
+
+    try:
+        existing_version_id = _find_existing_custom_application_source_version(
+            endpoint=endpoint,
+            token=token,
+            custom_application_source_id=custom_application_source_id,
+            version_token=version_token,
+        )
+        return existing_version_id
+    except KeyError:
+        upload_data: List[Tuple[str, Any]] = []
+        with contextlib.ExitStack() as stack:
+            if folder_path:
+                for root_path, _, file_paths in os.walk(folder_path):
+                    for path in file_paths:
+                        file_path = os.path.join(root_path, path)
+                        file = stack.enter_context(open(file_path, "rb"))
+                        upload_data.append(("file", (os.path.basename(file_path), file)))
+                        upload_data.append(("filePath", os.path.relpath(file_path, folder_path)))
+
+            if runtime_parameter_values_objs:
+                upload_data.append(
+                    (
+                        "runtimeParameterValues",
+                        json.dumps(
+                            [
+                                {camelize(k): v for k, v in param.to_dict().items()}
+                                for param in runtime_parameter_values_objs
+                            ]
+                        ),
+                    )
+                )
+
+            if "replicas" in kwargs:
+                upload_data.append(("replicas", str(kwargs["replicas"])))
+
+            if base_environment_id := kwargs.get("base_environment_id"):
```

Comment:
i think this is actually a different conditional; you want to test if "base_environment_id" is in kwargs which is not strictly equivalent to whether get returns a none-Falsey value; better to be explicit IMO; same to the below test on base env version id

Reply:
I don't understand the concern

Reply:
```
if "base_environment_id" in kwargs:
    upload_data.append(("baseEnvironmentId", kwargs['base_environment_id'))
```
Is both more clear/readable on first glance and also doesn't cause the reader/reviewer to start thinking about corner cases and whether they apply/need to be managed (e.g. what if user explicitly passes in None or a falsey value for this kwarg); the above basically says if user explicitly passed in base_environment_id, pass through what they requested; the original has the walrus operator complication and also makes the reader think about how does the get() work in concert with the walrus if falsey values are explicitly passed in? what if nothing was explicitly passed in? etc.